src/serac/physics/utilities/variational_form/integral.hpp

```diff
@@ -0,0 +1,640 @@
+  }
+  if constexpr (space::components > 1) {
+    return detail::Load<space::ndof, space::components>(u, e);
+  }
+};
+
+/**
+ * @brief Adds the contributions of the local residual to the global residual
+ * @param[inout] r_global The full element-decomposed residual
+ * @param[in] r_local The contributions to a residual from a single element
+ * @param[in] e The index of the element whose residual is @a r_local
+ */
+template <int ndof>
+void Add(const mfem::DeviceTensor<2, double>& r_global, tensor<double, ndof> r_local, int e)
+{
+  for (int i = 0; i < ndof; i++) {
+    r_global(i, e) += r_local[i];
+  }
+}
+/**
+ * @overload
+ * @note Used when each node has multiple DOFs
+ */
+template <int ndof, int components>
+void Add(const mfem::DeviceTensor<3, double>& r_global, tensor<double, ndof, components> r_local, int e)
+{
+  for (int i = 0; i < ndof; i++) {
+    for (int j = 0; j < components; j++) {
+      r_global(i, j, e) += r_local[i][j];
+    }
+  }
+}
+
+/**
+ * @brief Computes the arguments to be passed into the q-function (shape function evaluations)
+ * By default:
+ *  H1 family elements will compute {value, gradient}
+ *  Hcurl family elements will compute {value, curl}
+ *  TODO: Hdiv family elements will compute {value, divergence}
+ *  TODO: L2 family elements will compute value
+ *
+ * In the future, the user will be able to override these defaults
+ * to omit unused components (e.g. specify that they only need the gradient)
+ *
+ * @param[in] u The DOF values for the element
+ * @param[in] xi The position of the quadrature point in reference space
+ * @param[in] J The Jacobian of the element transformation at the quadrature point
+ * @tparam element_type The type of the element (used to determine the family)
+ */
+template <typename element_type, typename T, int dim>
+auto Preprocess(T u, const tensor<double, dim> xi, const tensor<double, dim, dim> J)
+{
+  // TODO: Static assert that element_type is an element??
+  if constexpr (element_type::family == Family::H1) {
+    return std::tuple{dot(u, element_type::shape_functions(xi)),
+                      dot(u, dot(element_type::shape_function_gradients(xi), inv(J)))};
+  }
+
+  if constexpr (element_type::family == Family::HCURL) {
+    // HCURL shape functions undergo a covariant Piola transformation when going
+    // from parent element to physical element
+    auto value = dot(u, dot(element_type::shape_functions(xi), inv(J)));
+    auto curl  = dot(u, element_type::shape_function_curl(xi) / det(J));
+    if constexpr (dim == 3) {
+      curl = dot(curl, transpose(J));
+    }
+    return std::tuple{value, curl};
+  }
+}
+
+/**
+ * @overload
+ * @note This specialization of detail::Preprocess is called when doing integrals
+ * where the spatial dimension is different from the dimension of the element geometry
+ * (i.e. surface integrals in 3D space, line integrals in 2D space, etc)
+ *
+ * QUESTION: are gradients useful in these cases or not?
```

Comment:
Per a discussion with @samuelpmishLLNL earlier, these would eventually be useful for reference loading finite deformation (e.g. finite deformation of a pressurized vessel).